STREAMS-1504 - Install Streams using an external Kafka cluster (#53)
Axway/streams-open-docs: content/en/docs/install/_index.md

### MariaDB settings

> The MariaDB database is automatically installed by default on your K8s cluster with the Streams Helm chart. But using an external one is recommended for production.

Reviewer: ```suggestion
By default, an embedded MariaDB database is installed on your K8s cluster next to Streams. For production, we recommend that you use an externalized one instead.
```

* or specify `--set mariadb.enabled=false` in the Helm Chart installation command.

> Then, according to your choice, configure your [externalized MariaDB](#externalized-mariadb-configuration) or your [automatically installed MariaDB](#embedded-mariadb-configuration).

Reviewer 2: ```suggestion
Then, according to your choice, configure your [externalized MariaDB](#externalized-mariadb-configuration) or your [embedded MariaDB](#embedded-mariadb-configuration).
```

  rootUsername: "my-streams-db-root-user"
```

##### Externalized MariaDB passwords

Passwords are required for Streams microservices to securely connect to Mariadb.

```sh
export NAMESPACE="my-namespace"
export MARIADB_ROOT_PASSWORD="my-mariadb-root-password"
export MARIADB_PASSWORD="my-mariadb-user-password"

kubectl create secret generic streams-database-passwords-secret --from-literal=mariadb-root-password=${MARIADB_ROOT_PASSWORD} --from-literal=mariadb-password=${MARIADB_PASSWORD} -n ${NAMESPACE}
```

##### Externalized MariaDB TLS

> For security purpose, it's highly recommended to enable TLS communication between your database and Streams microservices. You can enable [One-Way TLS](https://mariadb.com/kb/en/securing-connections-for-client-and-server/#enabling-one-way-tls-for-mariadb-clients) or [Two-Way TLS](https://mariadb.com/kb/en/securing-connections-for-client-and-server/#enabling-two-way-tls-for-mariadb-clients).

Reviewer 2: ```suggestion
For security purposes, it's highly recommended to enable TLS communication between your database and Streams microservices. You can enable [One-Way TLS](https://mariadb.com/kb/en/securing-connections-for-client-and-server/#enabling-one-way-tls-for-mariadb-clients) or [Two-Way TLS](https://mariadb.com/kb/en/securing-connections-for-client-and-server/#enabling-two-way-tls-for-mariadb-clients).
```

Not recommended for production.
{{< /alert >}}

### Kafka settings

By default, an embedded Kafka cluster is installed on your K8s cluster next to Streams. For production, we recommend that you use an externalized one instead.

To disable the Kafka installation, you can specify `--set kafka.enabled=false` in the Helm Chart installation command.

> Then, according to your choice, configure your [externalized Kafka](#externalized-kafka-configuration) or your [automatically installed Kafka](#embedded-kafka-configuration).

Reviewer 2: ```suggestion
Then, according to your choice, configure your [externalized Kafka](#externalized-kafka-configuration) or your [embedded Kafka](#embedded-kafka-configuration).
```

#### Externalized Kafka configuration

You must provide information to the Streams installation. You should specify `--set externalKafka.bootstrapServers="my.kafka.broker.1:port\,my.broker.2:port[...]"` in the Helm Chart installation command.

##### Externalized Kafka security settings

> For security purpose, it’s highly recommended to enable [SASL authentication](https://docs.confluent.io/current/kafka/authentication_sasl/index.html#authentication-with-sasl) and [TLS encryption](https://docs.confluent.io/current/kafka/encryption.html#encryption-with-ssl) for Kafka clients and brokers. You can enable both or neither.

Reviewer 2: ```suggestion
For security purposes, it’s highly recommended to enable [SASL authentication](https://docs.confluent.io/current/kafka/authentication_sasl/index.html#authentication-with-sasl) and [TLS encryption](https://docs.confluent.io/current/kafka/encryption.html#encryption-with-ssl) for Kafka clients and brokers. You can enable both or neither.
```

{{< alert title="Note" >}}

> Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enable or none of the two.

Reviewer 2: ```suggestion
Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enabled or none of the two.
```

{{< /alert >}}

> According to you choice, you must:

Reviewer 2: ```suggestion
According to your choice, you must:
```

* For SCRAM and TLS enabled:
    * Provide the user password using a k8s secret:
    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_USER_PASSWORD="my-kafka-password"

    kubectl create secret generic streams-kafka-passwords-secret --from-literal=client-passwords=${KAFKA_USER_PASSWORD} -n ${NAMESPACE}
    ```

>     * Provide a jks containing the kafka TLS truststore and its password:

Reviewer 2: ```suggestion
    * Provide a JKS containing the Kafka TLS truststore and its password:
```

    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_JKS_PASSWORD="my-kafka-jks-password"
    export KAFKA_JKS_PATH="my-kafka-jks-path"

    kubectl create secret generic streams-kafka-client-jks-secret --from-file=kafka.truststore.jks=${KAFKA_JKS_PATH} --from-literal=jks-password=${KAFKA_JKS_PASSWORD} -n ${NAMESPACE}
    ```
    * Set the [Helm parameters](#helm-parameters) `externalKafka.auth.clientUsername` with your Kafka username.

* For security disabled:
    * Set the [Helm parameters](#helm-parameters) `externalKafka.auth.clientProtocol` to `plaintext`.

#### Embedded Kafka configuration

##### Embedded Kafka security settings

> For security purpose, it’s highly recommended to enable [SASL authentication](https://docs.confluent.io/current/kafka/authentication_sasl/index.html#authentication-with-sasl) and [TLS encryption](https://docs.confluent.io/current/kafka/encryption.html#encryption-with-ssl) for Kafka clients and brokers. You can enable both or neither.

Reviewer 2: ```suggestion
For security purposes, it’s highly recommended to enable [SASL authentication](https://docs.confluent.io/current/kafka/authentication_sasl/index.html#authentication-with-sasl) and [TLS encryption](https://docs.confluent.io/current/kafka/encryption.html#encryption-with-ssl) for Kafka clients and brokers. You can enable both or neither.
```

SASL and TLS are enabled by default. As there is no sensitive data in Zookeeper, the communications with Zookeeper are in plaintext without authentication.

{{< alert title="Note" >}}

> Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enable or none of the two.

Reviewer 2: ```suggestion
Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enable or neither of the two.
```

{{< /alert >}}

> According to you choice, you must:

Reviewer 2: ```suggestion
According to your choice, you must:
```

* For SCRAM and TLS enabled:
    * Provide Kafka credentials using a k8s secret:
    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_CLIENT_PASSWORD="my-kakfa-client-password"
    export KAFKA_INTERBROKER_PASSWORD="my-kakfa-interbroker-password"

    kubectl -n ${NAMESPACE} create secret generic streams-kafka-passwords-secret --from-literal="client-passwords=${KAFKA_CLIENT_PASSWORD}" --from-literal="inter-broker-password=${KAFKA_INTERBROKER_PASSWORD}" Z
    ```
    * In order to configure TLS encryption, you need to have a valid truststore and one certificate per broker.

>         * They must all be integrated into Java Key Stores (JKS) files. Be careful as each broker needs its own keystore and a dedicated CN name matching the Kafka pod hostname as described in [bitnami documentation](https://github.com/bitnami/charts/tree/master/bitnami/kafka#enable-security-for-kafka-and-zookeeper).

Reviewer 2: ```suggestion
        * They must all be integrated into Java Key Stores (JKS) files. Be careful, as each broker needs its own keystore and a dedicated CN name matching the Kafka pod hostname as described in [bitnami documentation](https://github.com/bitnami/charts/tree/master/bitnami/kafka#enable-security-for-kafka-and-zookeeper).
```

Not recommended for production.
{{< /alert >}}

### Kafka settings

A Kafka cluster is automatically installed by default on your K8s cluster with the Streams Helm chart. But using an external one is recommended in production.

To disable the Kafka installation, you can either:

* Edit the `values.yaml` file and set the `kafka.enabled` entry as follow:

```yaml
kafka:
  enabled: false
```

> * or specify `--set kafka.enabled=false` in the Helm Chart installation command.

Reviewer 3: make it one sentence

Then, according to your choice, configure your [external Kafka](#external-kafka-configuration) or your [automatically installed Kafka](#helm-chart-kafka-configuration).

#### External Kafka configuration

> You must provide information to the Streams installation. You should edit the values.yaml file and set the `externalKafka` entry as follow:

Reviewer 3: switch to --set fashion

```yaml
externalKafka:
  bootstrapServers: "my.kafka.broker.1:port,my.broker.2:port[...]"
```

##### External Kafka security settings

For security purpose, it’s highly recommended to enable [SASL authentication](https://docs.confluent.io/current/kafka/authentication_sasl/index.html#authentication-with-sasl) and [TLS encryption](https://docs.confluent.io/current/kafka/encryption.html#encryption-with-ssl) for Kafka clients and brokers. You can enable both or neither.

{{< alert title="Note" >}}

> Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enable or neither.

Reviewer 3: TLS enabled, or none of the two.

{{< /alert >}}

According to you choice, you must:

* For SCRAM and TLS enabled:

>     * Provide the user password using secret:

Reviewer 3: using a k8s secret:

    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_USER_PASSWORD="my-kafka-password"

    kubectl create secret generic streams-kafka-passwords-secret --from-literal=client-passwords=${KAFKA_USER_PASSWORD} -n ${NAMESPACE}
    ```
    * Provide a jks containing the kafka TLS truststore and its password:
    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_JKS_PASSWORD="my-kafka-jks-password"
    export KAFKA_JKS_PATH="my-kafka-jks-path"

    kubectl create secret generic streams-kafka-client-jks-secret --from-file=kafka.truststore.jks=${KAFKA_JKS_PATH} --from-literal=jks-password=${KAFKA_JKS_PASSWORD} -n ${NAMESPACE}
    ```
    * Set the [Helm parameters](#helm-parameters) `externalKafka.auth.clientUsername` with your Kafka username.

> * For no security:

Reviewer 3: For security disabled:

    * Set the [Helm parameters](#helm-parameters) `externalKafka.auth.clientProtocol` to `plaintext`.

#### Helm chart Kafka configuration

##### Helm chart Kafka security settings

For security purpose, it’s highly recommended to enable [SASL authentication](https://docs.confluent.io/current/kafka/authentication_sasl/index.html#authentication-with-sasl) and [TLS encryption](https://docs.confluent.io/current/kafka/encryption.html#encryption-with-ssl) for Kafka clients and brokers. You can enable both or neither.

SASL and TLS are enabled by default. As there is no sensitive data in Zookeeper, the communications with Zookeeper are in plaintext without authentication.

{{< alert title="Note" >}}

> Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enable or neither.

Reviewer 3: same as previous note

{{< /alert >}}

According to you choice, you must:

* For SCRAM and TLS enabled:

>     * Provide Kafka credentials using secret:

Reviewer 3: using a k8s secret:

        * We provide you with a script to help with truststore and keystore generation (based on bitnami's script that properly handles Kubernetes deployment). You can also use your own truststore/privatekey:
        ```sh
        cd tools
        ./kafka-generate-ssl.sh
        ```
        * Create a secret which contains all the previously generated files:
        ```sh
        export NAMESPACE="my-namespace"
        export KAFKA_SECRET_PASSWORD="my-kakfa-secret-password"
        kubectl -n ${NAMESPACE} create secret generic streams-kafka-client-jks-secret --from-file="./truststore/kafka.truststore.jks" --from-file=./keystore/kafka-0.keystore.jks --from-file=./keystore/kafka-1.keystore.jks --from-file=./keystore/kafka-2.keystore.jks --from-literal="jks-password=${KAFKA_SECRET_PASSWORD}"
        ```

* For no security:
    * Set the following [Helm parameters](#helm-parameters):
        * `kafka.auth.clientProtocol` to `plaintext`
        * `kafka.auth.interBrokerProtocol` to `plaintext`
        * `kafka.auth.jaas` to `null`
        * `kafka.extraEnvVars` to `null`

{{< alert title="Note" >}}

> Do not use all security disabled for production.

Reviewer 3: Disabling security is not recommended for production.

Not recommended for production.
{{< /alert >}}

### Kafka settings

> By default, a Kafka cluster is automatically installed on your K8s cluster with the Streams Helm chart. But using an external one is recommended for production.

Reviewer: ```suggestion
By default, an embedded Kafka cluster is installed on your K8s cluster next to Streams. For production, we recommend that you use an externalized one instead.
```

{{< alert title="Note" >}}
Currently, Streams works only with SASL/SCRAM authentication (using the SHA-512 hash functions) **and** TLS enable or none of the two.
{{< /alert >}}

According to you choice, you must:

* For SCRAM and TLS enabled:
    * Provide the user password using a k8s secret:
    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_USER_PASSWORD="my-kafka-password"

    kubectl create secret generic streams-kafka-passwords-secret --from-literal=client-passwords=${KAFKA_USER_PASSWORD} -n ${NAMESPACE}
    ```
    * Provide a jks containing the kafka TLS truststore and its password:
    ```sh
    export NAMESPACE="my-namespace"
    export KAFKA_JKS_PASSWORD="my-kafka-jks-password"
    export KAFKA_JKS_PATH="my-kafka-jks-path"

    kubectl create secret generic streams-kafka-client-jks-secret --from-file=kafka.truststore.jks=${KAFKA_JKS_PATH} --from-literal=jks-password=${KAFKA_JKS_PASSWORD} -n ${NAMESPACE}
    ```
    * Set the [Helm parameters](#helm-parameters) `externalKafka.auth.clientUsername` with your Kafka username.

* For security disabled:
    * Set the [Helm parameters](#helm-parameters) `externalKafka.auth.clientProtocol` to `plaintext`.

> #### Helm chart Kafka configuration

Reviewer: I got confused because of the section title,  expecting it to cover the configuration requirements for Kafka in our Helm chart but it fact this section covers configuration requirements when deploying Kafka on Kubernetes vs deploying Kafka outside the k8s cluster. We need to change the different titles to gain clarity.
Maybe External Kafka / Internal Kafka or Kafka outside Kubernetes / Kafka on Kubernetes
@lbadenhop Any thoughts ?

Author: Maybe Embedded Kafka / Externalized Kafka?

Reviewer 2: Maybe Kafka outside Kubernetes / Kafka inside Kubernetes

> 

Reviewer 4: @benjaminranc @allandenis38 
Should you replace
`
By default, a Kafka cluster is automatically installed on your K8s cluster with the Streams Helm chart. But using an external one is recommended for production.`
`
To disable the Kafka installation, you can specify...
`
With something like
`By default, an embedded Kafka cluster is installed on your K8s cluster next to Streams. For production, we recommend using an externalized one instead.`
`
To disable the embedded Kafka installation, you can specify ...
`
for clarity and consistency ?
(just a suggestion)
(if so, same thing for mariadb)

Author: I agree for adding "embedded"
For the rest, I don't really have an opinion, both solutions suit me

| Parameter                             | Description                         | Mandatory | Default value |
| ------------------------------------- | ----------------------------------- | --------- | ------------- |
| kafka.enabled                         | Kafka installed in K8s with the Helm chart. If set to false, the `externalKafka` parameter will be used | no | true |
| kafka.auth.clientProtocol             | Authentication protocol used by Kafka client (must be "sasl_tls" or "plaintext") | no | sasl_tls |
| kafka.auth.interBrokerProtocol        | Authentication protocol internaly used by Kafka broker (must be "sasl_tls" or "plaintext") | no | sasl_tls |
| kafka.metrics.jmx.enabled             | Activate metrics endpoint for Kafka | no        | false         |

> | externalKafka.auth.clientUsername     | Username of the external Mariadb used by Streams (Only used when `kafka.enabled` set to false) | no | streams |

Reviewer: Name of the parameter `externalKafka.auth.clientUsername` refers to kafka but description refers to Mariadb.

| externalMariadb.tls.enabled           | Externalized MariaDB tls enabled (Only used when `mariadb.enabled` set to false) | no | true |
| externalMariadb.tls.twoWay            | Externalized MariaDB Two-Way tls enabled (Only used when `mariadb.enabled` set to false) | no | true |

#### Kafka parameters

| Parameter                             | Description                         | Mandatory | Default value |
| ------------------------------------- | ----------------------------------- | --------- | ------------- |
| kafka.enabled                         | Kafka installed in K8s with the Helm chart. If set to false, the `externalKafka` parameter will be used | no | true |
| kafka.auth.clientProtocol             | Authentication protocol used by Kafka client (must be "sasl_tls" or "plaintext") | no | sasl_tls |
| kafka.auth.interBrokerProtocol        | Authentication protocol internaly used by Kafka broker (must be "sasl_tls" or "plaintext") | no | sasl_tls |
| kafka.metrics.jmx.enabled             | Activate metrics endpoint for Kafka | no        | false         |

> | externalKafka.auth.clientUsername     | Username of the externalized Kafka used by Streams (Only used when `kafka.enabled` set to false) | no | streams |

Reviewer 2: ```suggestion
| externalKafka.auth.clientUsername     | Username of the externalized Kafka used by Streams (only used when `kafka.enabled` set to false) | no | streams |
```
